src/plugins/__init__.py

```diff
@@ -33,3 +35,34 @@ def load_plugins(specs: List[str]) -> List[Type[MetadataPlugin]]:
         moduleobj = import_module(module)
         result.append(getattr(moduleobj, classname))
     return result
+
+
+class PluginManager:
+    def __init__(self, specs: List[str], db: Database) -> None:
+        self.plugin_classes = load_plugins(specs)
+
+        active_plugins = []
+        for plugin_class in self.plugin_classes:
+            plugin_name = plugin_class.get_name()
+            plugin_config = db.get_plugin_config(plugin_name)
+            try:
+                active_plugins.append(plugin_class(db, plugin_config))
+                logging.info("Loaded plugin %s", plugin_name)
+            except Exception:
+                logging.exception("Failed to load %s plugin", plugin_name)
+        self.active_plugins = active_plugins
+
+    def filter(self, orig_name: str) -> Optional[str]:
+        """ Runs all available filter plugins on the provided file.
+        Returns new file path, or None. User should call cleanup() later. """
+        current_path: Optional[str] = orig_name
+        for plugin in self.active_plugins:
+            if not plugin.is_filter:
+                continue
+            if current_path:
+                current_path = plugin.filter(orig_name, current_path)
```

Comment:
The semantics of this snippet is kinda, weird, maybe a early `break` or `return` is missing somewhere?
As far as i can tell, currently, `current_path` can become None and be passed to `plugin.filter`. At the same time it looks like the `plugin.filter` method expects a `str` and not `Optional[str]` (calling gzip.open on it and expecting a str in typehints) 🤔

Reply:
yeah, you're right :thinking: I wonder why typings didn't catch the issue

Reply:
oh wait, you're not right. When `current_path` becomes `None` we won't  call `.filter` again, because `if current_path` will return False.

I don't remember what the reasoning was for structuring the code that way. I think we can simplify and just return early when current_path becomes None.

Reply:
looks better :+1: